tests/test_mps.py

```diff
@@ -332,3 +336,188 @@ def test_circ_approx_explicit(circuit: Circuit) -> None:
+
+
+@pytest.mark.parametrize(
+    "circuit",
+    [
+        pytest.lazy_fixture("q2_x0"),  # type: ignore
+        pytest.lazy_fixture("q2_x1"),  # type: ignore
+        pytest.lazy_fixture("q2_v0"),  # type: ignore
+        pytest.lazy_fixture("q2_x0cx01"),  # type: ignore
+        pytest.lazy_fixture("q2_x1cx10x1"),  # type: ignore
+        pytest.lazy_fixture("q2_hadamard_test"),  # type: ignore
+        pytest.lazy_fixture("q2_lcu1"),  # type: ignore
+        pytest.lazy_fixture("q2_lcu2"),  # type: ignore
+        pytest.lazy_fixture("q2_lcu3"),  # type: ignore
+        pytest.lazy_fixture("q3_cx01cz12x1rx0"),  # type: ignore
+        pytest.lazy_fixture("q5_line_circ_30_layers"),  # type: ignore
+    ],
+)
+@pytest.mark.parametrize(
+    "observable",
+    [
+        QubitPauliString({Qubit(0): Pauli.Z}),
+        QubitPauliString({Qubit(1): Pauli.X}),
+        QubitPauliString({Qubit(0): Pauli.X, Qubit(1): Pauli.Z}),
+    ],
+)
+def test_expectation_value(circuit: Circuit, observable: QubitPauliString) -> None:
+    pauli_to_optype = {Pauli.Z: OpType.Z, Pauli.Y: OpType.Z, Pauli.X: OpType.X}
+
+    # Use pytket to generate the expected value of the observable
+    ket_circ = circuit.copy()
+    for q, o in observable.map.items():
+        ket_circ.add_gate(pauli_to_optype[o], [q])
+    ket_sv = ket_circ.get_statevector()
+
+    bra_sv = circuit.get_statevector()
+
+    expected_value = bra_sv.conj() @ ket_sv
+
+    # Simulate the circuit and obtain the expected value
+    with CuTensorNetHandle() as libhandle:
+        mps = simulate(libhandle, circuit, ContractionAlg.MPSxGate)
+        assert np.isclose(
+            mps.expected_value(observable), expected_value, atol=mps._atol
+        )
+
+
+@pytest.mark.parametrize(
+    "circuit",
+    [
+        pytest.lazy_fixture("q2_x1"),  # type: ignore
+        pytest.lazy_fixture("q2_x0cx01"),  # type: ignore
+        pytest.lazy_fixture("q2_v0cx01cx10"),  # type: ignore
+        pytest.lazy_fixture("q2_hadamard_test"),  # type: ignore
+        pytest.lazy_fixture("q2_lcu2"),  # type: ignore
+    ],
+)
+def test_sample_circ_2q(circuit: Circuit) -> None:
+    n_samples = 200
+
+    q0 = circuit.qubits[0]
+    q1 = circuit.qubits[1]
+
+    # Compute the probabilities of each outcome
+    p = dict()
+    for outcome in range(4):
+        p[outcome] = abs(circuit.get_statevector()[outcome]) ** 2
+
+    # Compute the samples
+    sample_dict = {0: 0, 1: 0, 2: 0, 3: 0}
+    with CuTensorNetHandle() as libhandle:
+        mps = simulate(libhandle, circuit, ContractionAlg.MPSxGate)
+
+        # Take samples measuring both qubits at once
+        for _ in range(n_samples):
+            outcome_dict = mps.sample()
+            outcome = outcome_dict[q0] * 2 + outcome_dict[q1]
+            sample_dict[outcome] += 1
+
+    # Check sample frequency consistent with theoretical probability
+    for outcome, count in sample_dict.items():
+        assert np.isclose(count / n_samples, p[outcome], atol=0.1)

```

Comment:
`atol` is rather large - is it because of stochastisity of the sampling?

Reply:
Yeah, it generally succeeds with smaller atol (say 0.01) but I got some fails every now and then, so I just want to make sure this doesn't ever give false negatives. We could get atol to be smaller but that'd need more samples to get a better resolution, and I'd rather not make the tests take much longer.

Besides, `atol=0.1` was already enough to catch a couple of bugs. In our tested circuits there's always a few bitstrings that have much higher probability than others, so checking for `atol=0.1` is enough to detect if these are where they should be.

Reply:
Intuitive figure now appearing in the Jupyter notebook, showing that sampling works as expected:

![image](https://github.com/CQCL/pytket-cutensornet/assets/104848389/4cce1893-1cc7-4528-a031-d6bd10e34ebe)